internal/datacoord/policy.go

```diff
@@ -12,128 +12,383 @@
 package datacoord
 
 import (
-	"crypto/rand"
-	"math"
-	"math/big"
+	"sort"
+	"strconv"
+	"time"
 
+	"github.com/golang/protobuf/proto"
+	"github.com/milvus-io/milvus/internal/kv"
 	"github.com/milvus-io/milvus/internal/log"
 	"github.com/milvus-io/milvus/internal/proto/datapb"
 	"go.uber.org/zap"
+	"stathat.com/c/consistent"
 )
 
-// data node register func, simple func wrapping policy
-type dataNodeRegisterPolicy func(cluster []*NodeInfo, session *NodeInfo, buffer []*datapb.ChannelStatus) ([]*NodeInfo, []*datapb.ChannelStatus)
+// RegisterPolicy decide the channels mapping after registering the nodeID
+type RegisterPolicy func(store ROChannelStore, nodeID int64) ChannelOpSet
 
-// test logic, register and do nothing
-var emptyRegister dataNodeRegisterPolicy = func(cluster []*NodeInfo, session *NodeInfo, buffer []*datapb.ChannelStatus) ([]*NodeInfo, []*datapb.ChannelStatus) {
-	return []*NodeInfo{session}, buffer
+// EmptyRegister do nothing
+func EmptyRegister(store ROChannelStore, nodeID int64) ChannelOpSet {
+	return nil
 }
 
-// assign existing buffered channels into newly registered data node session
-var registerAssignWithBuffer dataNodeRegisterPolicy = func(cluster []*NodeInfo, session *NodeInfo, buffer []*datapb.ChannelStatus) ([]*NodeInfo, []*datapb.ChannelStatus) {
-	node := session.Clone(AddChannels(buffer))
-	return []*NodeInfo{node}, []*datapb.ChannelStatus{}
-}
+// BufferChannelAssignPolicy assign buffer channels to new registered node
+func BufferChannelAssignPolicy(store ROChannelStore, nodeID int64) ChannelOpSet {
+	info := store.GetBufferChannelInfo()
```

Comment:
If 10 new node come up 1 by 1, only the first node get all channels?
Should we limit the number of channels to return?

Reply:
I think we should balance channels when a new node comes up or periodically